File: src/two_means/test/test_twomeans.c
#include "twomeans.h"

list *cluster_array;

void test_create_cluster_array(int *C, int *pn, int *pKmax) {
  int k;
  int n = *pn, Kmax = *pKmax;

  init_cluster_array(C, n, Kmax);

  for (k = 0; k < Kmax; k++)
    print_que(cluster_array[k]);

  free_cluster_array(Kmax);
}

void test_two_means(double *X, int *pp, int *c, int *pn, double *cnt1, double *cnt2)
{
  int p = *pp, n = *pn;

  list ic1 = make_dummy();
  list ic2 = make_dummy();

  two_means(X, p, c, n, cnt1, cnt2, ic1, ic2);

  print_que(ic1);
  print_que(ic2);

  free_que(ic1);
  free_que(ic2);
}

File: src/two_means/test/test_twomeans.R
## create share object using the following code in your bash
##  R CMD SHLIB -c test_twomeans.c twomeans.c list.c

dyn.load("test_twomeans.so")

X <- as.matrix(iris[,-5])
n <- nrow(X)
p <- ncol(X)

Kmax <- 2
m1 <- X[10, ]
m2 <- X[50,]

km <- kmeans(X, centers=rbind(m1, m2))

## test create cluster array
ret <- .C("test_create_cluster_array",
          C=as.integer(km$cluster),
          n=as.integer(n),
          Kmax=as.integer(Kmax))
## test two means

ret <- .C("test_two_means",
          X = as.double(t(X)),
          p = as.integer(p),
          c = as.integer(1:n-1),
          n = as.integer(n),
          cnt1 = as.double(m1),
          cnt2 = as.double(m2))

ret$cnt1 - km$centers[1,]
ret$cnt2 - km$centers[2,]
